### PR TITLE
Refactored the executor exception handling for cache

### PR DESCRIPTION
**Pull request by rishitesh**
## What changes were proposed in this pull request?

Refactored the executor exception handling for cache closed exception.

## How was this patch tested?
precheckin pending

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 1]
```diff
@@ -429,74 +429,72 @@ private[spark] class Executor(
...
+          }
+          setTaskFinishedAndClearInterruptStatus()
+          execBackend.statusUpdate(taskId, TaskState.FAILED, serializedTaskEndReason)
+
+          // Don't forcibly exit unless the exception was inherently fatal, to avoid
+          // stopping other tasks unnecessarily.
+          if (Utils.isFatalError(t)) {
+            if (!isLocal) {
+              Thread.getDefaultUncaughtExceptionHandler.
                   uncaughtException(Thread.currentThread(), t)
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Are these changes mostly reversion of the changes done earlier? What does the diff against stock spark's version looks like?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -429,74 +429,72 @@ private[spark] class Executor(
...
-            if (Utils.isFatalError(t)) {
-              if (!isLocal) {
-                Thread.getDefaultUncaughtExceptionHandler.
+          }
+          setTaskFinishedAndClearInterruptStatus()
+          execBackend.statusUpdate(taskId, TaskState.FAILED, serializedTaskEndReason)
+
+          // Don't forcibly exit unless the exception was inherently fatal, to avoid
+          // stopping other tasks unnecessarily.
+          if (Utils.isFatalError(t)) {
```
Comment (sumwale) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: I think this should be improved to use something like SystemFailure.isJVMFailureError. In particular this should not treat the following as fatal: StackOverflowError, direct buffer OOME, OOME from Spark layer. The check in Gem SystemFailure.isJVMFailureError looks more correct (perhaps Spark Utils.isFatalError can itself be changed or move to a method that can be overridden in SnappyExecutor).

**Review by sumwale — COMMENTED** [review]
See one query and comment about isFatalError. Looks good otherwise.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 1 above)
Comment (rishitesh) [reply core/src/main/scala/org/apache/spark/executor/Executor.scala]: @sumwale Yes, most of the changes are due to previous changes. The current diff will be two exception handling and two no -op methods added.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 2 above)
Comment (rishitesh) [reply core/src/main/scala/org/apache/spark/executor/Executor.scala]: Ok. Will change it.

**Review by sumwale — APPROVED** [review]

**Pull request merged**